src/models/image.ts

```diff
@@ -30,6 +31,44 @@ export class IIIFImage {
         return this.append(`full/pct:${pct}/0/default.${extension}`);
     }
 
+    public getScaledAndCroppedUrl(pct: number, x: number, y: number, width: number, height: number, extension = 'jpg') {
+        return this.append(`${x},${y},${width},${height}/pct:${pct}/0/default.${extension}`);
+    }
+
+    // Returns a server-optimized scale factor for the image, based on the manifest.
+    // Returns a percentage.
+    //
+    // We take the boundingBox into account - its width and height are the actual width and height we consult
+    // (the default is the entire image's width and height)
+    public getOptimizedScaleFactor(expectedWidth: number, expectedHeight: number, boundingBox?: BoundingBox): number {
+        if (!this.manifest) {
+            console.warn("Can't get the scale factor of an image without loading the manifest first");
+            return 100;
+        }
+
+        const width: number = boundingBox ? boundingBox.width : this.manifest.width;
+        const height: number = boundingBox ? boundingBox.height : this.manifest.height;
+
+        if (this.manifest.sizes) {
+            const widthFactor = width / this.manifest.width;
+            const heightFactor = height / this.manifest.height;
+
+            for (const resolution of this.manifest.sizes) {
+
+                if (resolution.width * widthFactor  >= expectedWidth) {
+                    return Math.ceil(100 * resolution.width / this.manifest.width);
+                }
+                if (resolution.height *  heightFactor >= expectedHeight) {
+                    return Math.ceil(100 * resolution.height / this.manifest.height);
+                }
+            }
+        }
+
+        // Fallback - round up to the nearest 5%
+        const realScale = 100 * Math.min(expectedWidth / width, expectedHeight / height);
```

Comment:
I like this.

Reply:
We need to do something if the manifest has no sizes...

Reply:
That's what you've done here, yes? If `this.manifest.sizes` doesn't exist we jump down here, and if we never hit line 59 or 62 (maybe if `this.manifest.sizes` is an empty array), we end up here.  I guess the only question is where to set sizing breakpoints; in this case you've gone with breakpoints in 5% increments.  Maybe a larger interval would be better (I am thinking about browser caching), I don't really know.